Docs extproc (#9073)
solo-io/gloo: docs/content/guides/traffic_management/extproc/header-manipulation.md

---
title: Header manipulation
weight: 40
description: Walk through an example for how to manipulate request headers by using an extProc server. 
---

Set up an external processing (extProc) server that manipulates request headers for a sample app.

{{% notice note %}}
External processing is an Enterprise-only feature. 
{{% /notice %}}

{{% notice warning %}}
Envoy's external processing filter is considered a work in progress and has an unknown security posture. Do not use this feature in production environments. 
{{% /notice %}}

1. Before you begin, install [Gloo Edge Enterprise]({{% versioned_link_path fromRoot="/installation/enterprise/" %}}) in your cluster. 

2. Set up the extProc server. This example uses a prebuilt extProc server that manipulates request and response headers based on instructions that are sent in an `instructions` header. 
   
   {{< tabs >}}
{{% tab %}}
```yaml
kubectl apply -f- <<EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ext-proc-grpc
spec:
  selector:
    matchLabels:
      app: ext-proc-grpc
  replicas: 1
  template:
    metadata:
      labels:
        app: ext-proc-grpc
    spec:
      containers:
        - name: ext-proc-grpc

>           image: gcr.io/solo-test-236622/ext-proc-example-basic-sink:0.0.2

Reviewer: is the `test` repo the official one we want for public docs?

Author: yes that is the one we use for the tap server as well. product said it was ok to use it

   ext-proc-grpc-59d44ddf76-42q2x   1/1     Running   0          24m
   ```
   
4. Edit the default `Settings` custom resource to enable extProc in Gloo Edge. 
   ```
   kubectl edit settings default -n gloo-system
   ```
   
   Add the following extProc settings to the `spec` section: 
   ```yaml
   extProc: 
     grpcService: 
       extProcServerRef: 
         name: default-ext-proc-grpc-4444
         namespace: gloo-system
     filterStage: 
       stage: AuthZStage
       predicate: After
     failureModeAllow: false
     allowModeOverride: false
     processingMode: 
       requestHeaderMode: SEND
       responseHeaderMode: SKIP
       requestBodyMode: BUFFERED
       responseBodyMode: NONE
       requestTrailerMode: SKIP
       responseTrailerMode: SKIP
   ```
   
   |Setting|Description|
   |--|--|
   |`grpcService`| The configuration of the external processing server that you created earlier.|
   |`grpcService.exProcServerRef.name`| The name of the upstream that was created for the extProc server.|
   |`grpcService.exProcServerRef.namespace`| The namespace of the upstream that was created for the extProc server.|

>    |`filterStage`|Where in the filter chain do you want to apply the external processing.|

Reviewer: ```suggestion
   |`filterStage`|Where in the filter chain you want to apply the external processing.|
```

   |`failureModeAllow`|Allow the extProc server to continue when an error is detected during external processing. If set to `true`, the extProc server continues. If set to `false`, external processing is stopped and an error is returned to the Envoy proxy. |
   |`allowModeOverride`|Allow the extProc server to override the processing mode settings that you set. Default value is `false`. |

>    |`processingMode`|Decide how you want the extProc server to process request and response information. |

Reviewer: is there a list of supported modes we could provide?

Author: the processing modes are different depending on the field. I added possible values further below.

   |`processingMode.requestHeaderMode`|Send (`SEND`) or skip sending (`SKIP`) request header information to the extProc server. |
   |`processingMode.responseHeaderMode`|Send (`SEND`) or skip sending (`SKIP`) response header information to the extProc server. |

>    |`processingMode.requestBodyMode` </br>`processingMode.requestBodyMode`|Decide how to provide body information from the request or response to the extProc server. Available settings are <ul><li> `NONE`: Do not send the body. This is the default value. </li><li> `STREAMED`: Stream the body to the extProc server in pieces as they arrive at the proxy. </li><li> `BUFFERED`: Buffer the body in memory and send the entire body at once. If the body exceeds the maximum buffer limit, an error is sent to the downstream service. </li><li> `BUFFERED_PARTIAL`: Buffer the body in memory and send the entire body at once. If the body exceeds the buffer limit, the body content is sent in pieces up to the buffer limit. </li></ul>|

Reviewer: should the second `processingMode.requestBodyMode` be `processingMode.responseBodyMode`?

>    |`processingMode.requestBodyMode` </br>`processingMode.requestBodyMode`|Decide how to provide body information from the request or response to the extProc server. Available settings are <ul><li> `NONE`: Do not send the body. This is the default value. </li><li> `STREAMED`: Stream the body to the extProc server in pieces as they arrive at the proxy. </li><li> `BUFFERED`: Buffer the body in memory and send the entire body at once. If the body exceeds the maximum buffer limit, an error is sent to the downstream service. </li><li> `BUFFERED_PARTIAL`: Buffer the body in memory and send the entire body at once. If the body exceeds the buffer limit, the body content is sent in pieces up to the buffer limit. </li></ul>|

Reviewer: maybe add a colon - `Available settings are: <ul>`

>    

Reviewer: should there be an entry for the trailer modes?

---
title: Header manipulation
weight: 40
description: Walk through an example for how to manipulate request headers by using an extProc server. 
---

Set up an external processing (extProc) server that manipulates request headers for a sample app.

{{% notice note %}}
External processing is an Enterprise-only feature. 
{{% /notice %}}

{{% notice warning %}}

> Envoy's external processing filter is considered a work in progress and has an unknown security posture. Do not use this feature in production environments. For more information, see the [Envoy documentation](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_proc_filter#external-processing).

Reviewer 2: same suggestion as above
```suggestion
Envoy's external processing filter is considered a work in progress and has an unknown security posture. Use caution when using this feature in production environments. For more information, see the [Envoy documentation](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_proc_filter#external-processing).
```

{{% /notice %}}

1. Before you begin, install [Gloo Edge Enterprise]({{% versioned_link_path fromRoot="/installation/enterprise/" %}}) in your cluster. 

2. Set up the extProc server. This example uses a prebuilt extProc server that manipulates request and response headers based on instructions that are sent in an `instructions` header. 
   
   {{< tabs >}}
{{% tab %}}
```yaml
kubectl apply -f- <<EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ext-proc-grpc
spec:
  selector:
    matchLabels:
      app: ext-proc-grpc
  replicas: 1
  template:
    metadata:
      labels:
        app: ext-proc-grpc
    spec:
      containers:
        - name: ext-proc-grpc
          image: gcr.io/solo-test-236622/ext-proc-example-basic-sink:0.0.2
          imagePullPolicy: IfNotPresent
          ports:
            - containerPort: 18080
---
apiVersion: v1
kind: Service
metadata:
  name: ext-proc-grpc
  labels:
    app: ext-proc-grpc
  annotations:
    gloo.solo.io/h2_service: "true"
spec:
  ports:
  - port: 4444
    targetPort: 18080
    protocol: TCP
  selector:
    app: ext-proc-grpc
EOF
```
{{% /tab %}}
   {{< /tabs >}}

>    The `instruction` headers must be provided as a JSON string in the following format: 

Reviewer 2: ```suggestion
   The `instructions` header must be provided as a JSON string in the following format: 
```

  
   ```json 
   {
     "addHeaders": {
       "header1": "value1",
       "header2": "value2"
     },
     "removeHeaders": [ "header3", "header4" ],
     "setBody": "This is the new body",
     "setTrailers": {
       "trailer1": "value1",
       "trailer2": "value2"

>      }

Reviewer 2: let's remove the mention of body and trailers since they are not used/tested

   }
   ```
   
3. Verify that the extProc server is up and running. 
   ```sh
   kubectl get pods 
   ```
   
   Example output: 
   ```
   NAME                             READY   STATUS    RESTARTS   AGE
   ext-proc-grpc-59d44ddf76-42q2x   1/1     Running   0          24m
   ```
   

> 4. Edit the default `Settings` custom resource to enable extProc in Gloo Edge. 

Reviewer 2: this is fine but it may be a smoother experience if they just pass in the settings yaml via helm values when they install GE in Step 1

   ```
   kubectl edit settings default -n gloo-system
   ```
   
   Add the following extProc settings to the `spec` section: 
   ```yaml
   extProc: 
     grpcService: 
       extProcServerRef: 
         name: default-ext-proc-grpc-4444
         namespace: gloo-system
     filterStage: 
       stage: AuthZStage
       predicate: After
     failureModeAllow: false
     allowModeOverride: false
     processingMode: 
       requestHeaderMode: SEND
       responseHeaderMode: SKIP
       requestBodyMode: BUFFERED
       responseBodyMode: NONE
       requestTrailerMode: SKIP

>        responseTrailerMode: SKIP

Reviewer 2: we can remove body/trailer since we don't use them in the sample app
